### PR TITLE
Better parsing of AQL

### PR DESCRIPTION
**Pull request by stanistan**
function `split_on_comma` refactored to be generated by `explodeOnFn($delimiter)`

instead of `explode(' ')` for checking clauses (which is really unreliable), now using `splitOnWhitespace($str)`, also generated by `explodeOnFn()`

Corrects and issue where the order by was:

```
field = ''
```

and would be parsed as:

```
field = table.''
```

### CODE REVIEW

File: lib/core/functions.inc.php [hunk 1]
```diff
@@ -129,6 +129,81 @@ function shorten($text, $limit = 25) {
...
+	function explodeOnFn($delimiter) {
+		
+		// open => close (what matches what)
+		$closings = array(
+			'(' => ')',
+			"'" => "'",
+			'"' => '"'
+		);
+
+		return function($str) use($delimiter, $closings) {
```
Comment (will123195) [inline lib/core/functions.inc.php]: unnecessary nested function

File: lib/core/functions.inc.php [hunk 2]
```diff
@@ -129,6 +129,81 @@ function shorten($text, $limit = 25) {
 		return $text;
 	}
 
+	/**
+	 * makes a mini state machine that splits on the given $delimiter
```
Comment (will123195) [inline lib/core/functions.inc.php:134]: unclear why this is better than explode()

**Comment by stanistan** [comment]
It is true that this can be one function, `explodeOn($delimiter, $string)`, I think in this scenario, the function name will be more descriptive to what is happening. I'm open to suggestions.

Although I do think that `explodeOn()` should be added as a function as well.

Or maybe these functions should be renamed?

---

It's an improvement on explode for the purposes of AQL, and perhaps other string parsing (if it's necessary in the future), this example should clarify.

### Actual Test Output:

``` php
<?

$str = 'this, has ",commas and spaces" in weird places, that we would like to avoid';

// explode on space after exploding on comma
$exploded = array_map(function($r) {
    return explode(' ', $r);
}, explode(',', $str));

print_r($exploded);
/*
Array
(
    [0] => Array
        (
            [0] => this
        )

    [1] => Array
        (
            [0] => 
            [1] => has
            [2] => "
        )

    [2] => Array
        (
            [0] => commas
            [1] => and
            [2] => spaces"
            [3] => in
            [4] => weird
            [5] => places
        )

    [3] => Array
        (
            [0] => 
            [1] => that
            [2] => we
            [3] => would
            [4] => like
            [5] => to
            [6] => avoid
        )

)

*/

// using explodeOnFn -- ignores the $delimiter inside inside quotes/parenthesis

$exploded = array_map('explodeOnWhitespace', explodeOnComma($str));
print_r($exploded);

/*
Array
(
    [0] => Array
        (
            [0] => this
        )

    [1] => Array
        (
            [0] => has
            [1] => ",commas and spaces"
            [2] => in
            [3] => weird
            [4] => places
        )

    [2] => Array
        (
            [0] => that
            [1] => we
            [2] => would
            [3] => like
            [4] => to
            [5] => avoid
        )

)
*/
```

---

It splits on tokens and fixes many AQL parsing issues with nested statements and commas/quotes inside of quotes.

This ends up being really useful when splitting up fields in AQL as well as parsing clauses and the clause array. It fixes the issue that John was having (initial message in the pull request), because I can test for strings by checking if they start and end with a quote.

**Comment by stanistan** [comment]
Added renames, no longer returning a function.

**Pull request merged**